### PR TITLE
Priority streams

### PR DESCRIPTION
**Pull request by olisolomons**
Here are some changes that add a consumer that listens to multiple streams and processes messages in (mostly) priority order.

There is one important difference in functionality since you last saw it - a consumer now doesn't wait for the "abandoned messages check" to see if there are any messages that failed and are in its own pending queue - instead, when it is waiting for new messages and there aren't any (and there are no higher priority messages) it uses a single `XREADGROUP` call on all of the streams that might have pending messages, using ids of `0-0` - this returns messages for all streams that have pending messages all at the same time. Is also keeps track of which streams might have pending messages - since `gc-consumer-groups!` is gone, the only way a pending message can appear on a consumer group's backlog is if it was processed or it was `XCLAIM`ed from another stream, so we don't need to keep checking the same stream unless one of those two things happened on the stream.

There are a couple of possible issues with it:

1. as you can see in the tests, some of the messages aren't processed in priority order: when a message isn't processed (e.g. because of an exception in the callback function), it won't check for pending messages on that stream immediately - in the next iteration, it checks for new messages on the stream and when it doesn't find any, it moves onto the next stream and processes a lower priority message. It then checks for unprocessed messages on the higher priority stream and processes them. Is this ok?
2. If for some reason there are a lot of abandoned messages (e.g. from a dead consumer) or poison messages, only 100 (by default, configurable with `(assoc-in opts [:claim-opts :message-rescue-count] 123)`) messages are checked, so the backlog may increase over time. Should workers iterate over the entire backlog, or is this too much?

Anyway, let me know what you think!

### CODE REVIEW

**Comment by oliyh** [comment]
Hello,

Looks good, I'm happy with it - thank you again.

Please could you also update the README, and if you could write a short guide to help those upgrading from the current version that would be great as well.

Finally I don't know if you still have that sequence diagram but that could be useful too as documentation.

Cheers!

**Comment by olisolomons** [comment]
Hi,

I've (finally) done the README changes and diagram!

I re-did the diagram in graphviz, using gravizo.com to render it to an image. Would you prefer a different rending method? If you stick with this one, the url passed to gravizo.com would need to be changed to the new location of the file after merging.

Let me know what you think!

File: README.md [hunk 1]
```diff
@@ -59,40 +60,95 @@ Idempotent consumer group creation:
...
+
+Options to the consumer consist of:
+- `:block` ms to block waiting for a new message when there are no
+  pending messages on any of the streams
+- `:control-fn` a function for controlling the flow of operation, see `default-control-fn`
+- `:claim-opts` an options map for configuring how messages are
+  claimed from other consumers. See [Recovering from failures](#recovering-from-failures) for available options.
 
- Options to the consumer consist of:
+The behaviour of a single stream of a consumer that has multiple streams is given (slightly simplified) by this flowchart:
```
Comment (oliyh) [inline README.md:149]: "a single stream of a consumer"? Don't you just mean "The behaviour of a consumer that has multiple streams..."?

File: README.md [hunk 2]
```diff
@@ -59,40 +60,95 @@ Idempotent consumer group creation:
...
+     cleared.
+
+
+When checking for pending messages, if it has been sufficiently long
+since the last check, it will check for idle messages on the backlog
+of other consumers and claim them, or putting messages on the dlq if
+they have been retried too many times. This ensures that even if a
+consumer dies, its messages will still be processed.
+
+A consumer can also be passed multiple streams:
```
Comment (oliyh) [inline README.md:119]: Worth a new section heading here I think - "Multiple streams" or similar

**Comment by oliyh** [comment]
Hi, I've left some feedback for the README but it's looking really good.

Regarding the SVG I suggest you just check the SVG into the project under `/doc` and link to it from the SVG, name it the same as the graphviz file and it will be obvious that they go together.

Thanks

**Review by oliyh — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (olisolomons) [reply README.md:149]: I mean that the "current state" (the two ellipses at the top of the diagram) represents the state as-recorded for a single stream. It isn't always looks at pending messages for all streams or looking at new messages for all streams, but can be a mixture of the two. Not sure how to make that clearer though...

Comment (oliyh) [reply README.md:149]: How about "Each stream being processed by a multi-stream consumer will be processed as shown in this flowchart"?

**Comment by olisolomons** [comment]
Thanks for looking @oliyh, I've made the requested changes. How is it now?

**Review by oliyh — APPROVED** [review]

**Pull request merged**